### PR TITLE
actor critic loss

### PR DESCRIPTION
**Pull request by Haichao-Zhang**
This is a PR for https://github.com/HorizonRobotics/alf/issues/343

### CODE REVIEW

File: alf/algorithms/actor_critic_loss.py [hunk 1]
```diff
@@ -114,34 +117,33 @@ def __call__(self, training_info: TrainingInfo, value):
...
         def _summary():
             with self.name_scope:
-                tf.summary.scalar("values", tf.reduce_mean(value))
-                tf.summary.scalar("returns", tf.reduce_mean(returns))
-                tf.summary.scalar("advantages/mean",
-                                  tf.reduce_mean(advantages))
-                tf.summary.histogram("advantages/value", advantages)
-                tf.summary.scalar("explained_variance_of_return_by_value",
-                                  common.explained_variance(value, returns))
+                writer.add_scalar('values', value.mean())
```
[the remark this answers is not in the document]
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_loss.py]: Change to the centralized way using alf interface.

File: alf/algorithms/actor_critic_loss.py [hunk 2]
```diff
@@ -113,35 +113,35 @@ def __call__(self, training_info: TrainingInfo, value):
...
+            alf.summary.scalar(self._name + '/values', value.mean())
+            alf.summary.scalar(self._name + "/returns", returns.mean())
+            alf.summary.scalar(self._name + "/advantages/mean",
+                               advantages.mean())
+            alf.summary.histogram(self._name + "/advantages/value", advantages)
+            alf.summary.histogram(
+                self._name + "/explained_variance_of_return_by_value",
+                common.explained_variance(value, returns))
 
         if self._debug_summaries:
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_loss.py]:
```python
if self._debug_summaries and common.should_record_summaries():
```

File: alf/utils/dist_utils.py [hunk 3]
```diff
@@ -71,74 +73,47 @@ def estimated_entropy(dist: tfp.distributions.Distribution,
...
-        situations where its value is needed, and entropy_for_gradient where
-        you need to calculate the gradient of entropy.
+        entropy
+    """
+
+    def _compute_entropy(dist: torch.distributions.Distribution):
+        entropy = dist.entropy()
+        return entropy
+
+    entropies = list(map(_compute_entropy, nest.flatten(distributions)))
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: alf.nest.map_structure

File: alf/utils/dist_utils.py [hunk 4]
```diff
@@ -71,74 +73,47 @@ def estimated_entropy(dist: tfp.distributions.Distribution,
...
-    entropies = [e for e, eg in entropies]
-
-    return tf.add_n(entropies), tf.add_n(entropies_for_gradient)
+
+    def _compute_log_prob(single_distribution, single_action):
+        single_log_prob = single_distribution.log_prob(single_action)
+        return single_log_prob
+
+    nest.assert_same_structure(distributions, actions)
+    log_probs = [
```
Comment (emailweixu) [inline alf/utils/dist_utils.py]: alf.nest.map_structure

File: alf/algorithms/actor_critic_loss.py [hunk 5]
```diff
@@ -113,35 +113,35 @@ def __call__(self, training_info: TrainingInfo, value):
...
-                tf.summary.scalar("explained_variance_of_return_by_value",
-                                  common.explained_variance(value, returns))
+            alf.summary.scalar(self._name + '/values', value.mean())
+            alf.summary.scalar(self._name + "/returns", returns.mean())
+            alf.summary.scalar(self._name + "/advantages/mean",
+                               advantages.mean())
+            alf.summary.histogram(self._name + "/advantages/value", advantages)
+            alf.summary.histogram(
+                self._name + "/explained_variance_of_return_by_value",
+                common.explained_variance(value, returns))
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_loss.py]: common.explained_variance

File: alf/algorithms/actor_critic_loss.py [hunk 6]
```diff
@@ -151,9 +151,8 @@ def _summary():
...
 
     def _pg_loss(self, training_info, advantages):
-        action_log_prob = tfa_common.log_probability(
-            training_info.info.action_distribution, training_info.action,
-            self._action_spec)
+        action_log_prob = dist_utils.compute_log_probability(
+            training_info.info.action_distribution, training_info.action)
         return -advantages * action_log_prob
 
     def _calc_returns_and_advantages(self, training_info, value):
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_loss.py]: this function also needs to be implemented

File: alf/algorithms/actor_critic_loss_test.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+
+class TestLogProbability(unittest.TestCase):
+    def test_log_probability_(self):
+        M = td.categorical.Categorical(
+            torch.Tensor([[0.25, 0.75], [0.5, 0.5], [0.75, 0.25]]))
+        actions = torch.Tensor([0]).repeat(3)
+        expected = torch.Tensor([-1.38629436, -0.6931471, -0.287682])
+        obtained = compute_log_probability(M, actions)
+        np.testing.assert_array_almost_equal(expected, obtained)
```
Comment (emailweixu) [inline alf/algorithms/actor_critic_loss_test.py]:
```python
if __name__ == '__main__':
   unittest.main()
```

Comment (emailweixu) [reply alf/algorithms/actor_critic_loss_test.py]: also needs to add this test to `.ci-cd/build.sh`

File: alf/examples/ac_cart_pole.gin [hunk 8]
```diff
@@ -47,3 +47,4 @@ TrainerConfig.eval_interval=5
 TrainerConfig.debug_summaries=False
 TrainerConfig.summarize_grads_and_vars=False
 TrainerConfig.summary_interval=5
+TrainerConfig.use_tf_functions=False
```
Comment (emailweixu) [inline alf/examples/ac_cart_pole.gin]: undo this change

File: alf/algorithms/actor_critic_loss.py (hunk 2 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_loss.py]: changed

File: alf/utils/dist_utils.py (hunk 3 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: changed to using map_structure

File: alf/utils/dist_utils.py (hunk 4 above)
Comment (Haichao-Zhang) [reply alf/utils/dist_utils.py]: changed to using map_structure

File: alf/algorithms/actor_critic_loss.py (hunk 5 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_loss.py]: implemented

File: alf/algorithms/actor_critic_loss.py (hunk 6 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_loss.py]: supporting functions implemented

File: alf/algorithms/actor_critic_loss_test.py (hunk 7 above)
Comment (Haichao-Zhang) [reply alf/algorithms/actor_critic_loss_test.py]: test added

File: alf/examples/ac_cart_pole.gin (hunk 8 above)
Comment (Haichao-Zhang) [reply alf/examples/ac_cart_pole.gin]: removed

File: alf/utils/losses.py [hunk 9]
```diff
@@ -13,34 +13,31 @@
...
-        y (tf.Tensor): prediction
+        x (Tensor): label
+        y (Tensor): prediction
     Returns:
-        loss (tf.Tensor)
+        loss (Tensor)
     """
-    return tf.compat.v1.losses.huber_loss(
-        x, y, reduction=tf.compat.v1.losses.Reduction.NONE)
+    return torch.nn.L1Loss(reduction=None)
```
Comment (emailweixu) [inline alf/utils/losses.py]: huber loss is not L1Loss. It's SmoothL1Loss

File: alf/utils/value_ops.py [hunk 10]
```diff
@@ -143,43 +138,39 @@ def discounted_return(rewards, values, step_types, discounts, time_major=True):
...
         acc_discounted_value = acc_discounted_reward * discount + reward
         return is_last * value + (1 - is_last) * acc_discounted_value
 
-    returns = tf.scan(
-        fn=discounted_return_fn,
-        elems=(rewards, values, is_lasts, discounts),
-        reverse=True,
-        initializer=final_value,
-        back_prop=False)
+    returns[-1] = final_value
```
Comment (emailweixu) [inline alf/utils/value_ops.py]: You can avoid line 169 by this:
```python
returns = torch.zeros(rewards.shape, dtype=rewards.dtype)
ret = final_value
for t in reversed(range(rewards.shape[0])):
    ret = discounted_return_fn(returns[t + 1], rewards[t],
         values[t], is_lasts[t], discounts[t])
    returns[t] = ret
```

Comment (Haichao-Zhang) [reply alf/utils/value_ops.py]: Changed. Also did similar changes to normalized advantage estimation.

File: alf/utils/losses.py (hunk 9 above)
Comment (Haichao-Zhang) [reply alf/utils/losses.py]: Changed.
